lib/event-libs/sdevent/sdevent.c

```diff
@@ -0,0 +1,400 @@
+    struct lws *wsi = (struct lws*) userdata;
+    struct lws_context *context = wsi->a.context;
+    struct lws_context_per_thread *pt = &context->pt[(int)wsi->tsi];
+    struct lws_pollfd eventfd;
+
+    lws_context_lock(pt->context, __func__);
+    lws_pt_lock(pt, __func__);
+
+    if (pt->is_destroyed)
+        goto bail;
+
+    eventfd.fd = fd;
+    eventfd.events = 0;
+    eventfd.revents = 0;
+
+    // TODO handle revents error bits
+
+    if (revents & EPOLLIN) {
+        eventfd.events |= LWS_POLLIN;
+        eventfd.revents |= LWS_POLLIN;
+    }
+
+    if (revents & EPOLLOUT) {
+        eventfd.events |= LWS_POLLOUT;
+        eventfd.revents |= LWS_POLLOUT;
+    }
+
+    lws_pt_unlock(pt);
+    lws_context_unlock(pt->context);
+
+    lws_service_fd_tsi(context, &eventfd, wsi->tsi);
+
+    if (pt->destroy_self) {
+        lws_context_destroy(pt->context);
+        return -1;
+    }
+
+    // fire idle handler
+    struct sd_event_source *idletimer = pt_to_priv_sd(pt)->idletimer;
+    if (idletimer) {
+        sd_event_source_set_time(idletimer, (uint64_t) 0);
+        sd_event_source_set_enabled(idletimer, SD_EVENT_ON);
+    }
+
+    // allow further events
+    // Note: do not move the assignment up, lws_service_fd_tsi may invalidate it!
+    struct sd_event_source *watcher = wsi_to_priv_sd(wsi)->source;
+    if (watcher) {
+        sd_event_source_set_enabled(watcher, SD_EVENT_ONESHOT);
+    }
+
+    return 0;
+
+bail:
+    lws_pt_unlock(pt);
+    lws_context_unlock(pt->context);
+
+    return -1;
+}
+
+static void io_sd(struct lws *wsi, int flags) {
+
+    struct lws_context_per_thread *pt = &wsi->a.context->pt[(int)wsi->tsi];
+
+    // only manipulate if there is an event source and if the pt is still alive
+    if (!pt_to_priv_sd(pt)->io_loop || !wsi_to_priv_sd(wsi)->source || pt->is_destroyed) {
+        return;
+    }
+
+    // assert that the requested flags do not contain anything unexpected
+    if (!((flags & (LWS_EV_START | LWS_EV_STOP)) &&
+          (flags & (LWS_EV_READ | LWS_EV_WRITE)))) {
+        lwsl_err("%s: assert: flags %d", __func__, flags);
+        assert(0);
+    }
+
+    // we are overdoing a bit here, so it resembles the structure in libuv.c
+    if (flags & LWS_EV_START) {
+        if (flags & LWS_EV_WRITE)
+            wsi_to_priv_sd(wsi)->events |= EPOLLOUT;
+
+        if (flags & LWS_EV_READ)
+            wsi_to_priv_sd(wsi)->events |= EPOLLIN;
+
+        sd_event_source_set_io_events(wsi_to_priv_sd(wsi)->source, wsi_to_priv_sd(wsi)->events);
+        sd_event_source_set_enabled(wsi_to_priv_sd(wsi)->source, SD_EVENT_ONESHOT);
+    } else {
+        if (flags & LWS_EV_WRITE)
+            wsi_to_priv_sd(wsi)->events &= ~EPOLLOUT;
+
+        if (flags & LWS_EV_READ)
+            wsi_to_priv_sd(wsi)->events &= ~EPOLLIN;
+
+        sd_event_source_set_io_events(wsi_to_priv_sd(wsi)->source, wsi_to_priv_sd(wsi)->events);
+
+        if (!(wsi_to_priv_sd(wsi)->events & (EPOLLIN | EPOLLOUT)))
```

Comment:
Switch if/else clause to make this more understandable ( no `!` needed). Besides, is this even correct? If I understand correctly we do disable the event if at least one of the two event flags are set?